lib/saml/user_attributes/dslogon.rb

```diff
@@ -0,0 +1,99 @@
+# frozen_string_literal: true
+require 'saml/user_attributes/base_decorator'
+
+# TODO: remove these nocov comments when this is able to be tested.
+# :nocov:
+module SAML
+  module UserAttributes
+    class DSLogon < BaseDecorator
+      PREMIUM_ASSURANCE_LEVELS = %w(2 3).freeze
+
+      def uuid
+        attributes['uuid']
+      end
+
+      def email
+        attributes['email']
+      end
+
+      def dslogon_edipi
+        attributes['dslogon_uuid']
+      end
+
+      def dslogon_status
+        attributes['dslogon_status']
+      end
+
+      def gender
+        attributes['dslogon_gender']&.chars&.first&.upcase
+      end
+
+      def dslogon_deceased
+        attributes['dslogon_deceased']
+      end
+
+      def birth_date
+        attributes['dslogon_birth_date']
+      end
+
+      def first_name
+        attributes['dslogon_fname']
+      end
+
+      def middle_name
+        attributes['dslogon_mname']
+      end
+
+      def last_name
+        attributes['dslogon_lname']
+      end
+
+      def dslogon_idtype
+        attributes['dslogon_idtype']
+      end
+
+      def ssn
+        attributes['dslogon_idvalue']
+      end
+
+      def dslogon_assurance
+        attributes['dslogon_assurance']
+      end
+
+      def loa
+        { current: loa_current, highest: loa_highest }
+      end
+
+      def multifactor
+        attributes['multifactor']
+      end
+
+      private
+
+      # The first ones are values needed to query MVI
+      # The second ones are additional values that should override MVI (EDIPI or match)
+      # In short we might find that a user has inconsistencies in MVI with the EDIPI provided.
+      # The last part are ID.me specific attributes used by vets.gov
+      def serializable_attributes
+        %i(
+          first_name middle_name last_name email gender ssn birth_date
+          dslogon_edipi dslogon_status dslogon_deceased
+          uuid multifactor loa
+        )
+      end
+
+      # if the account has dslogon assurance 2 or 3 then the user has identity proofed
+      def loa_current
+        PREMIUM_ASSURANCE_LEVELS.include?(dslogon_assurance) ? 2 : 1
+      end
+
+      # if the account has dslogon assurance 2 or 3 there is no option to FICAM level up the account,
```

Comment:
As discussed on Slack, this implements very different semantics for loa_highest between this class and the idme user attributes class.

Reply:
I strongly think this should instead change to `loa: {"current": x, "highest": y, "available": z }`
The current loa_highest implementations in dslogon/mhv would populated "available". Dunno how to populate highest (= "highest achieved") for dslogon/mhv, if they similarly get a SAML attribute from id.me about the highest achieved.

Reply:
I am happy to make that change, it's just that "verify" and "level-up" are essentially the same thing. If "level-up" is necessary id.me does it and returns a leveled up state. If it's already "leveled-up" then they just return the leveled up state. So it shouldn't make any difference. The real problem is that the F/E is incorrectly relying on LOA state != 3, instead they should compare loa_current < loa_highest.

Reply:
What value does "highest achieved" have? The front-end doesn't need to know this. Shouldn't need to know this. All they should need to know is that certain services are not available, maybe if I loa_current < loa_highest and I level up to loa_highest I will have access... so they do that should get access. I think it's incorrect to put this in the hands of client to determine.

Reply:
I can try to see if ID.me is able to provide us with their level of access. In which case we'll go back to that logic, but I still think it's just a matter of semantics.